File: perl/template_method.pl
package Foo;

sub new {
	my $class = shift;
	my $self = {};
	bless $self, $class;
	return $self;
}

sub algorithm {
	my $self = shift;
	my $a = shift;
	my $b = shift;
	print "Step1\n";							# step1: invariant
	print "Step2: ", $self->op($a, $b), "\n";	# step2: variantt
	print "Step3\n";							# step3: invariant

	sub op {
		die "This is an abstract method!\n";
	}
}

package FooAdd;

use base Foo;

sub op {
	my $self = shift;
	my ($a, $b) = @_;
	return $a + $b;
}

package FooMultiply;

use base Foo;

sub op {
	my $self = shift;
	my ($a, $b) = @_;
	return $a * $b;
}

package Main;

my $obj1 = new FooAdd;
$obj1->algorithm(0, 5);
my $obj2 = new FooMultiply;
$obj2->algorithm(0, 5);

File: c++/template_method.cpp
#include <iostream>

using namespace std;

class Foo {
public:
	void algorithm(int a, int b) {
		cout << "Step 1\n";
		cout << "Step 2 ";
		op(a, b);
		cout << "Step 3\n";
	}
	virtual void op(int a, int b) = 0;
};

class FooAdd: public Foo {
public:
	void op(int a, int b) {
		cout << a + b << endl;
	}
};

class FooMultiply: public Foo {
public:
	void op(int a, int b) {
		cout << a * b << endl;
	}
};

int main(void) {
	FooAdd obj1;
	obj1.algorithm(2, 3);

	FooMultiply obj2;
	obj2.algorithm(2, 3);

	return 0;
}
